tests/backends/base/test_creation.py

```diff
@@ -40,3 +41,29 @@ def test_custom_test_name_with_test_prefix(self):
         test_connection.settings_dict['TEST'] = {'NAME': test_name}
         signature = BaseDatabaseCreation(test_connection).test_db_signature()
         self.assertEqual(signature[3], test_name)
+
+
+@mock.patch.object(connection, 'ensure_connection', return_value=None)
+@mock.patch('django.core.management.commands.migrate.Command.handle', return_value=None)
+class TestDbCreationTests(SimpleTestCase):
+    def test_migrate_test_setting_false(self, mocked_migrate, mocked_ensure_connection):
+        creation = connection.creation_class(connection)
+        try:
+            saved_settings = connection.settings_dict
```

Comment:
This line can be outside the `try` block.

Reply:
I think it's fine to leave it inside a `try` block.

Reply:
Granted, I haven't seen your latest edits, but ...

I normally try to minimize unnecessary code inside a try block. I like to keep the try block focused on what can actually fail and what actually requires cleanup. This statement is neither of those.

If the statement `saved_settings = connection.settings_dict` fails (I suppose either due to an `AttributeError` or some low level internal error preventing an assignment) there is nothing to cleanup as the reference was never assigned. If such an error were to happen, the `finally` block will run. But as `saved_settings` was never assigned correctly, the `finally` block will raise an unhandled `UnboundLocalError` rather than restore the global environment. This is easy to verify with the following mutation (for demonstration purposes):

```diff
diff --git a/tests/backends/base/test_creation.py b/tests/backends/base/test_creation.py
index c5c9761a5f..2d61407566 100644
--- a/tests/backends/base/test_creation.py
+++ b/tests/backends/base/test_creation.py
@@ -48,6 +48,7 @@ class TestDbSignatureTests(SimpleTestCase):
 class TestDbCreationTests(SimpleTestCase):
     def test_migrate_test_setting_false(self, mocked_migrate, mocked_ensure_connection):
         creation = connection.creation_class(connection)
+        del connection.settings_dict
         try:
             saved_settings = connection.settings_dict
             connection.settings_dict['TEST']['MIGRATE'] = False
```

Running this cases the test to fail with:

```
======================================================================
ERROR: test_migrate_test_setting_false (backends.base.test_creation.TestDbCreationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jon/devel/django/tests/backends/base/test_creation.py", line 53, in test_migrate_test_setting_false
    saved_settings = connection.settings_dict
  File "/home/jon/devel/django/django/db/__init__.py", line 28, in __getattr__
    return getattr(connections[DEFAULT_DB_ALIAS], item)
AttributeError: 'DatabaseWrapper' object has no attribute 'settings_dict'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.8/unittest/mock.py", line 1342, in patched
    return func(*newargs, **newkeywargs)
  File "/home/jon/devel/django/tests/backends/base/test_creation.py", line 59, in test_migrate_test_setting_false
    connection.settings_dict = saved_settings
UnboundLocalError: local variable 'saved_settings' referenced before assignment
```

But I would expect the test to fail with the `AttributeError`, not the `UnboundLocalError`.

So, IMO, it makes more sense to let the `AttributeError` be raised.

Reply:
Agreed.